src/control/server/instance_drpc.go

```diff
@@ -84,84 +84,62 @@ func (srv *IOServerInstance) CallDrpc(module, method int32, body proto.Message)
-	} else {
-		resp := &mgmtpb.DaosResp{}
-		if err = proto.Unmarshal(dresp.Body, resp); err != nil {
-			outErr = errors.WithMessagef(err, "rank %s dRPC unmarshal failed",
-				&rank)
-		} else if resp.GetStatus() != 0 {
-			outErr = errors.Errorf("rank %s dRPC returned DER %d",
-				&rank, resp.GetStatus())
-		}
+		return system.NewMemberResult(rank,
+			errors.WithMessagef(err, "rank %s dRPC failed", &rank),
+			system.MemberStateErrored)
 	}
 
-	if outErr == nil {
-		state = tState
+	resp := &mgmtpb.DaosResp{}
+	if err = proto.Unmarshal(dresp.Body, resp); err != nil {
+		return system.NewMemberResult(rank,
+			errors.WithMessagef(err, "rank %s dRPC unmarshal failed", &rank),
+			system.MemberStateErrored)
+	}
+	if resp.GetStatus() != 0 {
+		return system.NewMemberResult(rank,
+			errors.Errorf("rank %s: %s", &rank, drpc.DaosStatus(resp.GetStatus()).Error()),
+			system.MemberStateErrored)
 	}
 
-	return system.NewMemberResult(rank, action, outErr, state)
+	return system.NewMemberResult(rank, nil, tState)
 }
 
-// dPing attempts dRPC request to given rank managed by instance and return
+// TryDrpc attempts dRPC request to given rank managed by instance and return
 // success or error from call result or timeout encapsulated in result.
-func (srv *IOServerInstance) dPing(ctx context.Context) *system.MemberResult {
+func (srv *IOServerInstance) TryDrpc(ctx context.Context, method int32) *system.MemberResult {
 	rank, err := srv.GetRank()
 	if err != nil {
 		return nil // no rank to return result for
 	}
 
-	if !srv.isReady() {
-		return system.NewMemberResult(rank, "ping", nil, system.MemberStateStopped)
-	}
-
-	resChan := make(chan *system.MemberResult)
-	go func() {
-		dresp, err := srv.CallDrpc(drpc.ModuleMgmt, drpc.MethodPingRank, nil)
-		resChan <- drespToMemberResult(rank, "ping", dresp, err, system.MemberStateReady)
-	}()
-
-	select {
-	case <-ctx.Done():
-		if ctx.Err() == context.DeadlineExceeded {
-			return system.NewMemberResult(rank, "ping", ctx.Err(),
-				system.MemberStateUnresponsive)
-		}
-		return nil // shutdown
-	case result := <-resChan:
-		return result
-	}
-}
-
-// dPrepShutdown attempts dRPC request to given rank managed by instance and return
-// success or error from call result or timeout encapsulated in result.
-func (srv *IOServerInstance) dPrepShutdown(ctx context.Context) *system.MemberResult {
-	rank, err := srv.GetRank()
-	if err != nil {
-		return nil // no rank to return result for
+	localState := srv.LocalState()
+	if localState != system.MemberStateReady {
+		// member not ready for dRPC comms
+		return system.NewMemberResult(rank, nil, localState)
 	}
 
-	if !srv.isReady() {
-		return system.NewMemberResult(rank, "prep shutdown", nil, system.MemberStateStopped)
+	var tgtState system.MemberState
+	switch method {
+	case drpc.MethodPingRank:
+		tgtState = system.MemberStateReady
+	case drpc.MethodPrepShutdown:
+		tgtState = system.MemberStateStopping
+	default:
+		panic("unrecognised fanout method")
```

Comment:
I don't think we should panic here. Would prefer to instead return early with a MemberResult indicating that the fanout method was invalid. It's true that this would most likely be due to programmer error, but on the off chance that we manage to ship code that allows this default case to be hit, we don't want the control plane server to crash.

e.g.
```go
        default:
                err := errors.Errorf("unrecognised fanout method %d", method)
                return system.NewMemberResult(rank, err, system.MemberStateErrored)
```

Reply:
thought about this, if for some reason there is an unexpected case here we don't want the software to continue running because something is obviously awry and could cause unexpected behaviour and potentially data loss. that's why I think it should panic

Reply:
I don't agree with this. Production code should not panic except in very narrowly-defined circumstances from which there is no recovery or mitigation (e.g. -ENOMEM). In this case, if we somehow allowed an unknown value to be specified as the method, the recovery would be to exit early with an error, leaving things in a defined state. Crashing the control plane here would just be confusing.

Reply:
addressing in https://jira.hpdd.intel.com/browse/DAOS-3413